### PR TITLE
AI Provider Settings

### PR DESCRIPTION
**Pull request by Caleb-T-Owens**
This PR introduces the ability to better configure the AI within GitButler, allowing for bring your own key configurations and choosing between different strength models.

App preferences whilst logged in
image

App preferences whilst logged out
image

App preferences choosing a bring your own key option
image

Updated project settings will always show the AI settings
image

### CODE REVIEW

File: gitbutler-ui/src/lib/components/BranchCard.svelte
Comment (mtsgrd) [inline gitbutler-ui/src/lib/components/BranchCard.svelte]: Are you intentionally using `then` or would this be better written with `await`?

File: gitbutler-ui/src/routes/settings/+page.svelte
Comment (mtsgrd) [inline gitbutler-ui/src/routes/settings/+page.svelte]: For sake of consistency with the rest of the codebase can we please use `==`?

**Comment by mtsgrd** [comment]
Looking really solid 💪

File: gitbutler-ui/src/lib/components/BranchCard.svelte
Comment (Caleb-T-Owens) [reply gitbutler-ui/src/lib/components/BranchCard.svelte]: When I modified that part of the function, it wasn't yet declared as `async` but now it is so I'll move it over; thanks for the spot 👍

**Comment by Caleb-T-Owens** [comment]
> Looking really solid 💪

Thanks for taking a look, I realise it turned into quite a large PR so I'd be happy to talk over any of the changes if you have any questions/thoughts about any choices

File: gitbutler-ui/src/lib/backend/aiProviders.ts
Comment (mtsgrd) [inline gitbutler-ui/src/lib/backend/aiProviders.ts]: I think it would be good to avoid calling constructors inside constructors, it will make it easier to test this code in the future. What if we made the OpenAI client a parameter?

File: gitbutler-ui/src/lib/backend/aiProviders.ts
Comment (mtsgrd) [inline gitbutler-ui/src/lib/backend/aiProviders.ts]: Make this a multiline comment?

**Comment by mtsgrd** [comment]
For things operating on git like `gitGetConfig`, could we integrate them into an existing service or create a new one? I think we want to stick with using dependancy injection to keep the code testable, even though we currently don't have front end tests.

Everything else looks great to me!

File: gitbutler-ui/src/lib/components/AISettings.svelte
Comment (mtsgrd) [inline gitbutler-ui/src/lib/components/AISettings.svelte:61]: How about moving these into an `onMount(async ()=> {})` so you can use `await` instead of `then`?

File: gitbutler-ui/src/lib/components/AISettings.svelte
Comment (mtsgrd) [inline gitbutler-ui/src/lib/components/AISettings.svelte]: Do these run on first initialization atm?

File: gitbutler-ui/src/lib/backend/summarizer.ts
Comment (mtsgrd) [inline gitbutler-ui/src/lib/backend/summarizer.ts]: There is still something I want to figure out here. If we were testing either of the components where this is used, how would we swap this out for a mock `GitConfig` object? To me it feels like we need an `AiService` created in one of the `layout.ts` files and passed down to `BranchCard` and `CommitDialog`.

**Comment by mtsgrd** [comment]
Great progress! 💪

File: gitbutler-ui/src/lib/components/AISettings.svelte
Comment (Caleb-T-Owens) [reply gitbutler-ui/src/lib/components/AISettings.svelte]: Yes. Is there a clever way to prevent this, perhaps with the onMount? I looked to see if there was a svelte-y way to track changes and the solutions I found suggested having a second variable you assigned the previous value to which seemed very cumbersome.

File: gitbutler-ui/src/lib/components/AISettings.svelte
Comment (Caleb-T-Owens) [reply gitbutler-ui/src/lib/components/AISettings.svelte:61]: I like this suggestion, will do 👍

File: gitbutler-ui/src/lib/backend/summarizer.ts
Comment (Caleb-T-Owens) [reply gitbutler-ui/src/lib/backend/summarizer.ts]: I'll have a go at adding something like that.

In terms of how to provide the service, I'm not a huge fan of drilling the props down. There doesn't seem to be much use of the setContext and getContext API, is there a particular reason for this (and would it be reasonable to provide the GitConfig and AIService this way)?

Also, is AiSomething preferred over AISomething?

Comment (mtsgrd) [reply gitbutler-ui/src/lib/backend/summarizer.ts]: Yeah I'm not a fan either, and in fact, we're going to move to `setContext` and `getContext`. The only obstacle is that we need to replace the `projectId` constructor params with a store.. or something. I will have a look at that this coming week!

> Also, is AiSomething preferred over AISomething

Yeah, but it's just my personal preference. I don't like having e.g. three capital letters in a row. 😏

File: gitbutler-ui/src/lib/components/AISettings.svelte
Comment (Caleb-T-Owens) [reply gitbutler-ui/src/lib/components/AISettings.svelte]: Thinking about the AI service and how I want to interact with the RXJS, I might end up structuring them in such a way where this isn't a problem

Comment (Caleb-T-Owens) [reply gitbutler-ui/src/lib/components/AISettings.svelte]: So, even after using a writable, I needed to add in an "initialized" check which I'm not super thrilled about https://github.com/gitbutlerapp/gitbutler/pull/3084/commits/206d6cce07edfc4bf4d082b440423b8ad0da7844.

I thought registering the subscriber after the set call would have done the trick.

Comment (mtsgrd) [reply gitbutler-ui/src/lib/components/AISettings.svelte]: I'll pull down your branch locally so I can have a better look.

**Comment by Caleb-T-Owens** [comment]
A musing I had was, the aiService is more or less the summarizer with the creation of the provider just brought inside the class.

I'm going to have a think about this, because we're currently not succeeding at making the provider easy to inject

**Comment by Caleb-T-Owens** [comment]
> A musing I had was, the aiService is more or less the summarizer with the creation of the provider/client just brought inside the class.
>
> I'm going to have a think about this, because we're currently not succeeding at making the provider easy to inject

I was concerned about how to test it partially because I've not tried vitest before apart from seeing someone use it on stream, so I thought I'd give it a go, and I've been able to reasonably cover most of the logic

**Comment by mtsgrd** [comment]
Just leaving a note that I'll review the new changes as soon as I get a chance!

**Comment by Caleb-T-Owens** [comment]
I've just seen the new context changes, I'll go rebase in a bit and now the userService can be gotten from the userContext, I'll have a go passing it into the `commit` and `branch` methods... which I think I'll rename to `summarizeCommit` and `summarizeBranch` because it would then read as `aiService.summarizeCommit`

**Comment by Caleb-T-Owens** [comment]
I've done the rebase. I'll do the tweaks I said I'll do in a little bit, but then I've got some other things I need to work on this weekend

**Comment by mtsgrd** [comment]
I just sat down to look at this again. Sorry for my absence, I was deep in the weeds trying to wrap my head around both streams and dependency injection. ✌️

**Comment by Caleb-T-Owens** [comment]
No worries @mtsgrd!

I've gone and done that refactor I mentioned.

Part of me is wanting to extract the prompt construction and post processing into their own functions to better test them. What are your thoughts on that?

**Comment by mtsgrd** [comment]
@Caleb-T-Owens hey, sorry, I pushed here even though I meant to push to my own branch. I didn't realize GitButler can push to other peoples' forks. :)

Anyway, feel free to force push if you want to take some suggestions but tweak it. I just wanted to get this pushed before I go have some beers tonight. 🥳

**Comment by Caleb-T-Owens** [comment]
@mtsgrd 

> hey, sorry, I pushed here even though I meant to push to my own branch. I didn't realize GitButler can push to other peoples' forks. :)
>
> Anyway, feel free to force push if you want to take some suggestions but tweak it. I just wanted to get this pushed before I go have some beers tonight. 🥳

No worries! GitHub has an option to permit maintainers to push to your fork which is pretty nifty. I can now pull them into my branch with the upstream commits card now too 💪!

Enjoy your evening!

---

I see you've changed https://github.com/gitbutlerapp/gitbutler/pull/3084/commits/1247e5253ad8d3d9cc130599686075f8a39e7e58#diff-c828d978a2adbaa41707e94fea99a318b081ed67c3c3fa2fb610416dbc4431bfR23-R24 from using null to undefined. I typed that using `null` because `invoke` seems to be returning `null` but it was more inconvenient due to the bind value of text boxes wanting to receive `string | undefined` so I'm happy to make the move over. I am going to add in a `?? undefined` to ensure our types are consistent with our actual values.

---

Unrelated to your changes, I was having a think about https://github.com/gitbutlerapp/gitbutler/pull/3084/files#diff-17085abea331547efcd2493dbd64eb700ea1f3108919330b2a6748eb56ab9060R30-R35 where we're saying `$: setContext(...)`.

To me this seems a little odd because if we're in the scenario where `data.something` changes and we set the state, none of the `getContext` calls will react to that as their components won't have re-rendered. Aside from that, I was also under the impression that the load() method is awaited before the `+layout.svelte` is rendered meaning that the items in the data object shouldn't be getting updated after the `+layout.svelte` is rendered. If that assumption is true, doesn't that mean that we don't need to react regardless?

Perhaps I'm missing some sveltey wizardry that is going on in the background that I don't know about, so I'm keen to hear about this

**Comment by mtsgrd** [comment]
Just quickly before I head out.
- Yeah let's convert the `null` to `undefined`, and pretend `null` barely even exists in this code base
- The services located under `/[projectId]/` get update in the corresponding `layout.ts` file, but changing projects re-renders everything with a `getContext` call, so we're cool for now.
- I would perhaps prefer if we only had global services rather than some that depend on `projectId`, but let's refactor that later

File: gitbutler-ui/src/lib/backend/aiService.test.ts
Comment (mtsgrd) [inline gitbutler-ui/src/lib/backend/aiService.test.ts]: Just realized we probably want to tie this to the new enum?

Comment (Caleb-T-Owens) [reply gitbutler-ui/src/lib/backend/aiService.test.ts]: Yes - I was going to do that 👍

File: gitbutler-ui/src/lib/backend/aiService.test.ts
Comment (mtsgrd) [inline gitbutler-ui/src/lib/backend/aiService.test.ts]: Could you still update these strings to point to `ConfigKeys`? On the same note, I didn't necessarily mean to leave it named what it is. Perhaps it would be better to rename it e.g. `GitConfigKey`?

**Comment by mtsgrd** [comment]
I'm pretty stoked how well written this feature has become. Feel free to land the changes at any time 👍

**Review by mtsgrd — APPROVED** [review]

File: gitbutler-ui/src/lib/components/Select.svelte
Comment (mtsgrd) [inline gitbutler-ui/src/lib/components/Select.svelte]: Does this perhaps fit on one row?

File: package.json
Comment (mtsgrd) [inline package.json:19]: Side note, I wonder if we don't have a formatting specified for json in ci since it doesn't complain? We should perhaps keep the original unless we're now doing tabs.

Comment (Caleb-T-Owens) [reply package.json:19]: The tauri.conf.*.json files are all tabs and the gitbutler-ui/package.json is also tabs, but, I'd be happy to put that package.json back if we're not ready to make a particular decision on that.

Comment (mtsgrd) [reply package.json:19]: You're right, we should prefer consistency with the other json files. :)

File: gitbutler-ui/src/lib/backend/aiService.test.ts
Comment (Caleb-T-Owens) [reply gitbutler-ui/src/lib/backend/aiService.test.ts]: Yep, I missed those strings, I'll go and do those.

As for the name, part of me wants to indicate that they're related to the AI functionality and go for `GitAIConfigKey` to differentiate between some other config key enum in the future.

Part of me still likes the idea of having getters and setters for the fields, perhaps now on the GitConfigService, so we can say `await gitConfigService.getOpenAIKeyOption()` (or maybe as a method declared as `get`) so we don't need to worry about being consistent with default values in other areas of the code

Comment (mtsgrd) [reply gitbutler-ui/src/lib/backend/aiService.test.ts]: >await gitConfigService.getOpenAIKeyOption()

How many getters and setters do you think we could eventually end up with?

Comment (Caleb-T-Owens) [reply gitbutler-ui/src/lib/backend/aiService.test.ts]: > How many getters and setters do you think we could eventually end up with?

At the minute we've got two wrappers for local storage (persisted.ts & userSettings.ts), the gitbutler.json, various data stored in the new TOML files, and the git config.

I think there are several considerations around reducing the quantity of locations we store data, whether we want to sync the users settings to the could, and deciding on if we want functionality to provide project level overrides like vscode.

In the long run, my thought is that it would be helpful to have one unified interface (even if behind the scenes it writes to different providers) for accessing reading and writing settings so we don't end up with duplicate functionality and inconsistent interfaces.

---

Now is certainly not the time to reinvent that particular wheel. However, by adding these getters and setters, we're going to completely avoid any key name problems, inconsistent default problems, and typing problems in application code. To me it seems like a very sensible thing to do given that the cost of removing this abstraction is going to be pretty much equivalent to finding and replacing the `get` and `set` methods.

Comment (Caleb-T-Owens) [reply gitbutler-ui/src/lib/backend/aiService.test.ts]: I've use the enum in the other tests and renamed it.

Perhaps refactoring into setters and getters can be left to another time.

Something I do want to do before we merge in, is create a better metric for when the AI generation button should be disabled, because currently you must be logged in if you're providing your own token, but apart from that, if we're happy from a UI perspective, I'd be happy for this to go ahead and be merged in

Comment (Caleb-T-Owens) [reply gitbutler-ui/src/lib/backend/aiService.test.ts]: I've made the change to allow using AI without a user if a key is provided and updated the settings to better reflect that

File: gitbutler-ui/src/lib/components/BranchCard.svelte
Comment (mtsgrd) [inline gitbutler-ui/src/lib/components/BranchCard.svelte]: Since there is no async calls in this onMount we don't need to pass it an async arrow function.

File: gitbutler-ui/src/lib/backend/aiService.ts
Comment (mtsgrd) [inline gitbutler-ui/src/lib/backend/aiService.ts]: I think this function name is hard to interpret, could it be better described as `validateConfiguration`?

File: gitbutler-ui/src/lib/backend/aiService.ts
Comment (mtsgrd) [inline gitbutler-ui/src/lib/backend/aiService.ts:118]: Can we rewrite these two checks without the object type `Boolean`? We only have primitive strings and primitive boolean types in this code base.

In the latter I would just remove the boolean, and in the former I would return `!!userToken`.

**Comment by mtsgrd** [comment]
How about we land this today and then iterate?

File: gitbutler-ui/src/lib/backend/aiService.ts
Comment (Caleb-T-Owens) [reply gitbutler-ui/src/lib/backend/aiService.ts]: Sure! I think I was going for the ruby pattern where we define proposition methods because `?` can be part of a name, so we'd be able to call `aiService.configurationValid?` but `validateConfiguration` or maybe doing `hasValidConfiguration` would be better.

Comment (Caleb-T-Owens) [reply gitbutler-ui/src/lib/backend/aiService.ts]: I also want to extract the large propositions into some intermediary constants to clarify the intent

File: gitbutler-ui/src/lib/components/BranchCard.svelte
Comment (Caleb-T-Owens) [reply gitbutler-ui/src/lib/components/BranchCard.svelte]: Thanks for spotting that. That is just something I missed when committing

**Comment by Leopere** [comment]
this is exciting but does it add ollama support?  I'd like either ollama or grok or something that I get more choice or privacy with preferrably.

**Comment by Caleb-T-Owens** [comment]
Hi @Leopere, this certainly sets the groundwork for adding in supporting custom models over a user provided API endpoint, but doesn't yet provide that option.

In my dev environment I hacked in a basic AIClient that talked to ollama (which took about 5 minutes to do), but it made it clear that there is a good portion of work that needs to be done, before these weaker models can be anywhere near useful.

I summarised what I deemed to be prerequisite on discord: https://discord.com/channels/1060193121130000425/1204524097241878629/1219776902978600970. (invite to the server if needed: https://discord.gg/B7rv6qBejb)

**Comment by Leopere** [comment]
Yeah it’s kind of a shame that we’re not quite close to parity in oss to accomplish onprem llms yet

**Comment by PavelLaptev** [comment]
@Caleb-T-Owens @mtsgrd if you do new updated and this involves UI — include me to reviewers

**Comment by PavelLaptev** [comment]
Otherwise, I'll change what you did afterward. Which is bad practice.

**Comment by Caleb-T-Owens** [comment]
@PavelLaptev will do 👍

**Comment by PavelLaptev** [comment]
@Caleb-T-Owens thanks 🙏

**Pull request merged**
